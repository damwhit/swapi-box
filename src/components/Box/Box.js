import React, { Component } from 'react';
import { BrowserRouter, Route } from "react-router-dom";
import FavoritesButton from '../FavoritesButton/FavoritesButton.js';
import CategoryButton from '../CategoryButton/CategoryButton.js';
import ResultsContainer from '../ResultsContainer/ResultsContainer.js';
import ApiHelper from '../ApiHelper/ApiHelper.js';
import './Box.css';

class Box extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      isLoaded: false,
      category: null,
      numFavorites: 0,
      resources: JSON.parse(localStorage.getItem('resources')) || [],
    };
  }

  componentDidMount() {
    this.countFavorites();
  }

  countFavorites() {
    this.setState({
      numFavorites: this.state.resources.filter(resource => resource.isFavorite).length
    });
  }

  changeCategory(category) {
    this.setState({
      category: category,
    });
    const resourceExists = this.state.resources.some((resource) => resource.category === category);
    if (resourceExists && category !== 'favorites') return;
    ApiHelper.fetchResources(category);
    // const { error, isLoaded } = this.state;
    // if (error) {
    //   return <div>Error: {error.message}</div>;
    // } else if (!isLoaded) {
    //   return <div>Loading...</div>;
    // } else {
    //   return <div>loaded</div>
    // }
  }

  toggleFavoriteResource = (name) => {
    const resources = this.state.resources;
    const resource = resources.find(resource => resource.name === name);
    resource.isFavorite = !resource.isFavorite;
    this.setState({ resources: resources });
    localStorage.setItem('resources', JSON.stringify(resources));
    this.countFavorites();
  }
  
  renderCategoryButton(category) {
    return (
      <CategoryButton 
        onClick={() => this.changeCategory(category)} 
        value={category}
      />
    );
  }

  render() {
    if (true) {
      return (
        <BrowserRouter>
          <main className="box">
            <section>
              <h1>SWAPI-Box</h1>
              <FavoritesButton 
                onClick={() => this.changeCategory('favorites')} 
                value={this.state.numFavorites}
              />
            </section>
            <section>
              {this.renderCategoryButton('people')}
              {this.renderCategoryButton('planets')}
              {this.renderCategoryButton('vehicles')}
            </section>
            <Route path='/:category' render={({ match }) => {
              return (
                <ResultsContainer 
                  category={match.params.category}
                  onClick={this.toggleFavoriteResource} 
                  resources={this.state.resources}
                />
              );
            }} />
          </main>
        </BrowserRouter>
      );
    }
  }
}

export default Box;
