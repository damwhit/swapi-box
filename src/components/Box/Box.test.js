import React from 'react';
import { mount, shallow } from 'enzyme';
import Box from './Box';

describe('Box', () => {
  it('should update the state with the number of favorite resources', () => {
    // Setup
    const wrapper = mount(<Box />);
    const mockResources = [
      { name: 'Han Solo', isFavorite: true },
      { name: 'Chewbacca', isFavorite: true },
      { name: 'Luke Skywalker', isFavorite: false },
    ];

    wrapper.setState({
      resources: mockResources
    });

    // Execution
    wrapper.instance().countFavorites();

    // Expectation
    expect(wrapper.state('numFavorites')).toEqual(2);
  })
});
