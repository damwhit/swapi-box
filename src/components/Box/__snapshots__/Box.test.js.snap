// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Box />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <main
        className="box"
      >
        <section>
          <h1>
            SWAPI-Box
          </h1>
          <FavoritesButton
            numFavorites={0}
            onClick={[Function]}
          />
        </section>
        <section>
          <CategoryButton
            category="people"
            onClick={[Function]}
          />
          <CategoryButton
            category="planets"
            onClick={[Function]}
          />
          <CategoryButton
            category="vehicles"
            onClick={[Function]}
          />
        </section>
        <Switch>
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />
          if
          <Route
            path="/:category(favorites|people|planets|vehicles)"
            render={[Function]}
          />
          <Route
            render={[Function]}
          />
        </Switch>
      </main>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <section>
            <h1>
              SWAPI-Box
            </h1>
            <FavoritesButton
              numFavorites={0}
              onClick={[Function]}
            />
          </section>,
          <section>
            <CategoryButton
              category="people"
              onClick={[Function]}
            />
            <CategoryButton
              category="planets"
              onClick={[Function]}
            />
            <CategoryButton
              category="vehicles"
              onClick={[Function]}
            />
          </section>,
          <Switch>
            <Route
              exact={true}
              path="/"
              render={[Function]}
            />
            if
            <Route
              path="/:category(favorites|people|planets|vehicles)"
              render={[Function]}
            />
            <Route
              render={[Function]}
            />
          </Switch>,
        ],
        "className": "box",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                SWAPI-Box
              </h1>,
              <FavoritesButton
                numFavorites={0}
                onClick={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "SWAPI-Box",
              },
              "ref": null,
              "rendered": "SWAPI-Box",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "numFavorites": 0,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "section",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CategoryButton
                category="people"
                onClick={[Function]}
              />,
              <CategoryButton
                category="planets"
                onClick={[Function]}
              />,
              <CategoryButton
                category="vehicles"
                onClick={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "category": "people",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "category": "planets",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "category": "vehicles",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "section",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Route
                exact={true}
                path="/"
                render={[Function]}
              />,
              "if",
              <Route
                path="/:category(favorites|people|planets|vehicles)"
                render={[Function]}
              />,
              <Route
                render={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "exact": true,
                "path": "/",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "if",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "path": "/:category(favorites|people|planets|vehicles)",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "main",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <main
          className="box"
        >
          <section>
            <h1>
              SWAPI-Box
            </h1>
            <FavoritesButton
              numFavorites={0}
              onClick={[Function]}
            />
          </section>
          <section>
            <CategoryButton
              category="people"
              onClick={[Function]}
            />
            <CategoryButton
              category="planets"
              onClick={[Function]}
            />
            <CategoryButton
              category="vehicles"
              onClick={[Function]}
            />
          </section>
          <Switch>
            <Route
              exact={true}
              path="/"
              render={[Function]}
            />
            if
            <Route
              path="/:category(favorites|people|planets|vehicles)"
              render={[Function]}
            />
            <Route
              render={[Function]}
            />
          </Switch>
        </main>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <section>
              <h1>
                SWAPI-Box
              </h1>
              <FavoritesButton
                numFavorites={0}
                onClick={[Function]}
              />
            </section>,
            <section>
              <CategoryButton
                category="people"
                onClick={[Function]}
              />
              <CategoryButton
                category="planets"
                onClick={[Function]}
              />
              <CategoryButton
                category="vehicles"
                onClick={[Function]}
              />
            </section>,
            <Switch>
              <Route
                exact={true}
                path="/"
                render={[Function]}
              />
              if
              <Route
                path="/:category(favorites|people|planets|vehicles)"
                render={[Function]}
              />
              <Route
                render={[Function]}
              />
            </Switch>,
          ],
          "className": "box",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h1>
                  SWAPI-Box
                </h1>,
                <FavoritesButton
                  numFavorites={0}
                  onClick={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "SWAPI-Box",
                },
                "ref": null,
                "rendered": "SWAPI-Box",
                "type": "h1",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "numFavorites": 0,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "section",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <CategoryButton
                  category="people"
                  onClick={[Function]}
                />,
                <CategoryButton
                  category="planets"
                  onClick={[Function]}
                />,
                <CategoryButton
                  category="vehicles"
                  onClick={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "category": "people",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "category": "planets",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "category": "vehicles",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "section",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Route
                  exact={true}
                  path="/"
                  render={[Function]}
                />,
                "if",
                <Route
                  path="/:category(favorites|people|planets|vehicles)"
                  render={[Function]}
                />,
                <Route
                  render={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "exact": true,
                  "path": "/",
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "if",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "path": "/:category(favorites|people|planets|vehicles)",
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": "main",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
